twoliter/src/project.rs

```diff
@@ -157,64 +155,85 @@ impl ImageName {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {
-        fmt::Display::fmt(&self.get(), f)
-    }
+/// This is used to `Deserialize` a project, then run validation code before returning a valid
+/// [`Project`]. This is necessary both because there is no post-deserialization serde hook for
+/// validation and, even if there was, we need to know the project directory path in order to check
+/// some things.
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct UnvalidatedProject {
+    schema_version: SchemaVersion<1>,
+    release_version: String,
+    sdk: Option<ImageName>,
+    toolchain: Option<ImageName>,
 }
 
-impl<const N: u32> Serialize for SchemaVersion<N> {
-    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_u32(self.get())
+impl UnvalidatedProject {
+    /// Constructs a [`Project`] from an [`UnvalidatedProject`] after validating fields.
+    async fn validate(self, path: impl AsRef<Path>) -> Result<Project> {
+        let filepath: PathBuf = path.as_ref().into();
+        let project_dir = filepath
+            .parent()
+            .context(format!(
+                "Unable to find the parent directory of '{}'",
+                filepath.display(),
+            ))?
+            .to_path_buf();
+
+        self.check_release_toml(&project_dir).await?;
+
+        Ok(Project {
+            filepath,
+            project_dir,
+            schema_version: self.schema_version,
+            release_version: self.release_version,
+            sdk: self.sdk,
+            toolchain: self.toolchain,
+        })
     }
-}
 
-impl<'de, const N: u32> Deserialize<'de> for SchemaVersion<N> {
-    fn deserialize<D>(deserializer: D) -> Result<SchemaVersion<N>, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        let value: u32 = Deserialize::deserialize(deserializer)?;
-        if value != Self::get_static() {
-            Err(Error::custom(format!(
-                "Incorrect project schema_version: got '{}', expected '{}'",
-                value,
-                Self::get_static()
-            )))
-        } else {
-            Ok(Self)
+    /// Issues a warning if `Release.toml` is found and, if so, ensures that it contains the same
+    /// version (i.e. `release-version`) as the `Twoliter.toml` project file.
+    async fn check_release_toml(&self, project_dir: &Path) -> Result<()> {
+        let path = project_dir.join("Release.toml");
+        if !path.exists() || !path.is_file() {
+            // There is no Release.toml file. This is a good thing!
+            trace!("This project does not have a Release.toml file (this is not a problem)");
+            return Ok(());
+        }
+        warn!(
+            "A Release.toml file was found. Release.toml is deprecated. Please remove it from \
+             your project."
+        );
```

Comment:
nit: this could encourage people building from the main repo to send a PR to remove `Release.toml` - might want to drop it.